app/src/main/java/com/example/myapplication/ui/LoginFragment.kt

```diff
@@ -1,16 +1,58 @@
 package com.example.myapplication.ui
 
+import android.content.Context
 import android.os.Bundle
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
+import android.widget.Toast
 import androidx.fragment.app.Fragment
+import androidx.lifecycle.ViewModelProvider
 import com.example.myapplication.databinding.FragmentLoginBinding
+import com.example.myapplication.vm.LoginViewModel
+
+class LoginFragment : Fragment() {
 
-class LoginFragment: Fragment() {
     private lateinit var binding: FragmentLoginBinding
-    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+
+    private lateinit var loginViewModel: LoginViewModel
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
         binding = FragmentLoginBinding.inflate(inflater, container, false)
+        loginViewModel = ViewModelProvider(this).get(LoginViewModel::class.java)
+        binding.lifecycleOwner = this
+        val emailEdit = binding.firstNameEdit
+        val passwordEdit = binding.lastNameEdit
+        binding.logInButton.setOnClickListener { loginViewModel.fieldsValuesValidation(emailEdit, passwordEdit) }
+        loginViewModel.retrieveSavedUser()
+
+
+        loginViewModel.emptyFieldsError.observe(viewLifecycleOwner) {
+            if (it == true) {
```

Comment:
You can do `if (it) { ... }`

Reply:
_it_ is an observable, so it can be null, so you have to check for true, false, or null D:

Reply:
That's true, in that case I would use the null safety operator to avoid the `== true`